Doc new (#582)
marrink-lab/vermouth-martinize: doc/source/file_formats.rst

      substitution patterns to be applied throughout the file being
      read.
    - Macro values are substituted using the name with a preceding $.
      This is similar to the use of variables in shell scripting and
      makes it easy to write generalizations and to use and change
      default values.
    - The following example specifies that the protein default
      backbone bead type is P2. It can be referred to in the following
      sections of the file as ``$prot_default_bb_type``
      .. code-block::

        [ macros ]
        prot_default_bb_type P2

- ``[ variables ]``
    - ``optional``
    - The variables section has no further subsections and lists a
      number of variable stored as key value pairs in the `force field
      object <data: force field>`. This allows retrieving the parameters
      using ``force_field.variables[key] = value``.
    - Variables are used to control force field wide parameters that
      are tied to a specific force field version.
    - For example, the text below specifies that the bond type of the
      elastic networkx should be 1 for the force field.
      .. code-block:: 

        [ variables ]

>         elastic_network_bond_type 1

Author: I think this is more a martinize2

- ``[ citations ]``

    - ``optional``; but note that martinize2 expects some citations

>       to be present.

Author: yes but when running martinize2 and selection certain options like elastic network or Go model.

>       to be present.

Reviewer: Yes exactly. That's why the citations.bib cannot be empty, but you don't need to have  a [citations] section. As it reads it's ambiguous which of the two is expected.

    - The citations section lists the citations to be used for the
      force field. Citations are named and refer to an entry in the
      bibtex file `citations.bib` in the force field directory.
    - Citations can also be specified for moleculetypes, links, and
      modifications, in a citation subsection.
    - An example of this is:
      .. code-block:: 

        [ citations ]

>         Martini3

Author: I don't understand why here? This section only deals with the citations directive. A more comprehensive examples come below or do you mean having an all encompassing example?

>         Martini3

Reviewer: I would expand the example a bit to show not only how to put FF-wide citations, but also citations specific for e.g. moleculetype. And yes, the moleculetype comes below, but if I'm adding a model and want to add the citations, I'd look for the answer here.

>         Martini3

Author: ok now I understand your comment; I'll update it

        cgsecstruc "C"

- ``[ modification ]``
    - ``optional``
    - Modifications can be used to edit molecules or parts thereof
      (blocks), e.g., for specifying protonation states. Each
      modification starts with a line with the name. Thereafter may
      follow subsections as under links. A modification may add,
      remove, or change atoms, interactions and/or edges, using the
      corresponding subsections.

Allowed sub-directives: Moleculetype
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
- ``[ atoms ]``
    - ``mandatory``
    - Each line in the atoms section describes one particle,
      corresponding to a node in the molecular graph. The description
      comprises the following fields:

        * atom number
        * atom type
        * residue index
        * residue name
        * atom name
        * charge group (optional)
        * charge (optional)
        * mass (optional)

    - An example is shown below:
      .. code-block:: 

        [ atoms ]
        ;id type resnr residu atom cgnr   charge mass
        1   P5   1     GLY    BB     1      0    47

- ``[ edges ]``
    - ``optional``
    - Edges will be added to the molecular graph when required through

>       interactions, but they can also be added explicitly by listing

Author: done note added and slightly rephrased

      of specific atoms. An attributes statement is a JSON style
      mapping of key/value pairs, similar to those used in the
      #meta syntax (see below).
    - Note that here the syntax is different from the moleculetype
      atoms directive. This directive requires the particle name
      followed by a dict of attributes e.g ``BB {"resname": "ALA"}``
    - To overwrite atom attributes of existing atoms when a link is
      applied the user can provide a dict of parameters within using
      the replace key as follows:
      .. code-block:: 

        [ atoms ]
        BB {"replace": {"charge": -1}}

- ``[ interaction_name ]``
    - ``optional``
    - A link may list any number of interactions to be added, if a link
      applies. The syntax is the same as for the `moleculetype` sub-
      directive. However, when the listed particles are not within the
      same residue a prefix has to be provided that specifies the order
      relative to a given residue. The following prefixes are allowed:

        * +, ++, +++ : 	first, second, third following residue
        * -, --, --- : 	first, second, third previous residue

>         * > :		other residue (not in order)

Author: Nope this comes straight from the old docs; let me investigate

>         * > :		other residue (not in order)

Author: yes this is a mistake probably based on the cystein bridge example.

- ``[ features]``
    - ``optional``
    - The features subsection lists features to apply to the link
      itself. These can be used to control the application of links
      during the building of topologies. For example, setting the
      feature 'scfix' will cause the links to be applied only if the
      option -scfix is given to martinize2.

- ``[ molmeta ]``
    - ``optional``
    - The molmeta subsection lists metadata to be added/changed in the
      molecular graph. These metadata can be used (and modified) by
      Vermouth's processors and for provenance.

- ``[ edges ]``
    - ``optional``
    - Within the context of links, the edges specify that an
      edge should or shouldn't be present, respectively, for the link
      pattern to match.
    - They should mostly be used in cases, where interactions are
      applied for which edges cannot be made automatically.

- ``[ non-edges ]``
    - ``optional``
    - This directive specifies that an edge should be absent in order
      for the link to apply. Note that the first particle must be
      present in the link and the second one must the partner
      with which to not form an edge.

>     - This syntax is likely to be deprecated in the near future.

Author: I would not do it because it is actually required for the new protein model and other models. Unfortunetly, I think we'll have to keep it

>     - This syntax is likely to be deprecated in the near future.

Reviewer: Probably one of us (or Paulo) should then have a look at the protein model files....

The issue is that non-edges are slooow

>     - This syntax is likely to be deprecated in the near future.

Author: ahh yes of course we could try to do this; but for now I'd say we leave it in. In the end we said that things might get deprecated

Allowed sub-directives: Modifications
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
- ``[ atoms ]``
    - ``mandatory``

>     - The atoms subsection under a modification lists both anchors and
>       atoms to be added to anchors or changed. Entries consist of an
>       atom name followed by an attributes statement. Atoms that are
>       added need to set the "PTM_atom" attribute to true and require
>       a valid "element" attribute. The "replace" attribute may be set
>       to a (nested) JSON dict, listing the atom attributes to be
>       changed and the new values corresponding to the modification.
>       Such changes can also be applied to atoms already present in the
>       molecular graph, i.e., the 'non-PTM atoms'.`

Reviewer: ```suggestion
    - The atoms subsection under a modification lists both anchors and
      atoms to be added to anchors or changed. Entries consist of an
      atom name followed by an attributes statement. Atoms that are
      added need to set the "PTM_atom" attribute to True and require
      a valid "element" attribute. Atoms for which the "PTM_atom" 
      attribute is absent (or False) must already be described by the 
      relevant :ref:`block <data.Block>` with the same atomname. 
      The "replace" attribute may be set to a (nested) JSON dict, listing
      the atom attributes to be changed and the new values corresponding
      to the modification. Such changes can also be applied to atoms
      already present in the molecular graph, i.e., the 'non-PTM atoms'.`
```

- ``[ interaction_name ]``
    - ``optional``
    - A link may list any number of interactions to be added, if a
      modification applies. The syntax is the same as for the
      link sub directive.

> Special meta data
> ^^^^^^^^^^^^^^^^^

Reviewer: ```suggestion
Special meta data
~~~~~~~~~~~~~~~~
```
Otherwise it becomes a subsection of the modification subsections. Also, check the number of ~, I eyeballed it.

Author: done; but was missing one ~

statement.

.. code-block:: 

    [ link ]
    #meta {"group": "Side chain bonds", "ifdef": "FLEXIBLE"}

Metadata can also be added to a single line by adding an attribute
statement as the last element.

Mapping files (.map & .mapping)
-----------------------------------------------
A mapping specifies the conversion from one force field description to
another. If the transformation is from a higher resolution force field
to a lower resolution, e.g., from the canonical description to
Martini, the process is typically called 'forward mapping'.

The vermouth library currently utilizes two mapping formats. The ``.map``
format, which was originally developed for the backward program, is
used to describe how two :ref:`blocks <data:block>` correspond to each
other. The second format (``.mapping``) is exclusively used in the
context of :ref:`modifications <data:modification>` and is an extension
to the first format.

File structure (.map)
~~~~~~~~~~~~~~~~~~~~~

The file is structured into sections, each beginning with a directive
enclosed in square brackets (``[]``).

Allowed directives .map
^^^^^^^^^^^^^^^^^^^^^^^
The format recognizes the following directives:

- ``[molecule]``

>     - This directive is immediately followed by a single line containing
>       an alphanumeric string specifying the residue name. This name
>       denotes the residue under consideration.

Reviewer: ```suggestion
    - This directive is immediately followed by a single line containing
      an alphanumeric string specifying the residue name. This name
      denotes the residue under consideration. A :ref:`block <data.Block>`
      with this name must be defined in both the ``[from]`` and ``[to]`` force fields.
```

    - ``mandatory``

- ``[from]``
    - The directive is followed by a single line containing an
      alphanumeric string corresponding to name of the origin
      (i.e. higher resolution) force field.
    - ``mandatory``

- ``[to]``
    - The directive is followed by a single line containing an
      alphanumeric string corresponding to the name of the target
      (i.e. higher resolution) force field.
    - ``mandatory``

- ``[martini]``
    - The directive is followed by any number of lines. Each line must
      contain space separated bead names.

>     - ``mandatory``
> - ``[atoms]``

Reviewer: ```suggestion
    - ``mandatory``

- ``[atoms]``
```

>     - This directive introduces a section that can span multiple lines.
>       Each line within this section must adhere to the following
>       format:
> 
>       - An integer specifying the atom number.
>       - An alphanumeric string corresponding to an atom name in the
>         origin force field.
>       - Any number of bead names. These beads must have been previously
>         listed under the ``[martini]`` directive.
>     - ``mandatory``

Reviewer: ```suggestion
    - This directive introduces a section that can span multiple lines.
      Each line within this section must adhere to the following
      format:

      - An integer specifying the atom number.
      - An alphanumeric string corresponding to an atom name in the
        origin force field.
      - Any number of bead names. These beads must have been previously
        listed under the ``[martini]`` directive.
    - All atoms described by the referenced block should be described.
    - ``mandatory``
```

>     - ``mandatory``
> - ``[chiral]``
>     - Contains chirality specifications used for the original backwards
>       program.
>     - ``ignored``
> - ``[trans]``
>     - Contains geometry specifications used for the original backwards
>       program.
>     - ``ignored``
> - ``[out]``
>     - Contains geometry specifications used in the original backwards
>       program.
>     - ``ignored``

Reviewer: ```suggestion
    - ``mandatory``

- ``[chiral]``
    - Contains chirality specifications used for the original backwards
      program.
    - ``ignored``

- ``[trans]``
    - Contains geometry specifications used for the original backwards
      program.
    - ``ignored``

- ``[out]``
    - Contains geometry specifications used in the original backwards
      program.
    - ``ignored``
```

Example of .map file
^^^^^^^^^^^^^^^^^^^^

.. code-block:: 

    [ molecule ]
    ALA ALA
    [ martini ]
    BB SC1
    [ atoms ]
     1     N    BB
     2    HN    BB
     3    CA    BB
     5    CB    SC1
     9     C    BB
    10     O    BB

File structure (.mapping)
~~~~~~~~~~~~~~~~~~~~~~~~~
The file is structured into sections, each beginning with a directive
enclosed in square brackets (``[]``).

Allowed directives .mapping
^^^^^^^^^^^^^^^^^^^^^^^^^^^

- ``[modification]``
    - Marks the beginning of a modification block.
      This directive does not require any following content.

>     - ``mandatory``

Reviewer: Note that this is only a half-truth. .mapping files also work for Blocks (and parts of blocks), and can do some fancy things such as bridging residue boundaries.We can probably leave that out though.

Author: agreed! Let's keep that for us now

- ``[from]``
    - Followed by the name of the origin force-field (e.g., amber).
    - ``mandatory``

- ``[to]``
    - Followed by the name of the target force-field (e.g., martini3001).
    - ``mandatory``

> - ``[from blocks]`` and ``[to blocks]``
>     - Each followed by an alphanumeric string describing a block name
>       that needs to be modified by the modification at hand.
>     - ``mandatory``
> 
> - ``[from nodes]``
>     - Lists all nodes present in the modification description.
>     - ``mandatory``
> 
> - ``[from edges]``
>     - Contains all edges between the to-be-modified block and the
>       modification.
>     - ``mandatory``

Reviewer: ```suggestion
- ``[from blocks]`` and ``[to blocks]``
    - Each followed by the name of the modification in the respective
      force fields
    - ``mandatory``

- ``[from nodes]``
    - Lists all nodes that should be part of the mapping that are not
      yet described by ``[from block]``
    - ``optional``

- ``[from edges]``
    - Contains all edges that are part of the mapped fragment that are
      not described by ``[from block]``. In particular, all edges concerning
      nodes in ``[from nodes]`` must be listed here.
    - ``optional``
```
